File: sdk/ruby/srift.rb
# SRIFT Ruby SDK — zero-dep client. Ruby 2.7+ / 3.x / JRuby / TruffleRuby.
#
#   require_relative 'srift'
#   s = Srift::Client.new
#   r = s.quick_share('/abs/path/file.zip')
#   puts r['shareUrl']

require 'net/http'
require 'json'
require 'uri'

module Srift
  class Error < StandardError; end

  class Client
    DEFAULT_BASE = 'http://127.0.0.1:3822'

    def initialize(base_url: nil, timeout: 30)
      @base = (base_url || ENV['SRIFT_BASE_URL'] || DEFAULT_BASE).chomp('/')
      @timeout = timeout
    end

    def status;             get('/status');                end
    def state;              get('/state');                 end
    def chat_history;       get('/chat/history');          end

    def quick_share(file_path, session_name: nil)
      post('/quick-share', filePath: file_path, sessionName: session_name)
    end
    def start_session(name: nil, room_secret: nil)
      post('/session/start', sessionName: name, roomSecret: room_secret)
    end
    def join_session(session_id, username: nil, room_secret: nil)
      post('/session/join', sessionId: session_id, username: username, roomSecret: room_secret)
    end
    def approve_join(temp_user_id); post('/session/approve', tempUserId: temp_user_id); end
    def reject_join(temp_user_id, reason: nil); post('/session/reject', tempUserId: temp_user_id, reason: reason); end
    def kick_user(user_id); post('/session/kick', userId: user_id); end
    def close_session;      post('/session/close');                  end
    def send_file(file_path, protocol: nil); post('/send', filePath: file_path, protocol: protocol); end
    def accept_transfer(file_id, save_dir: nil); post('/receive', fileId: file_id, saveDir: save_dir); end
    def send_chat(message); post('/chat/send', message: message); end
    def list_transfers
      status['activeTransfers'] || []
    end

    private

    def get(path);  do_request(Net::HTTP::Get.new(URI(@base + path))); end
    def post(path, body = {})
      req = Net::HTTP::Post.new(URI(@base + path), 'Content-Type' => 'application/json')
      req.body = JSON.generate(body.compact)
      do_request(req)
    end

    def do_request(req)
      uri = req.uri
      res = Net::HTTP.start(uri.host, uri.port, read_timeout: @timeout) { |h| h.request(req) }
      raise Error, JSON.parse(res.body)['error'] || "HTTP #{res.code}" if res.code.to_i >= 400
      res.body.empty? ? {} : JSON.parse(res.body)
    rescue Errno::ECONNREFUSED => e
      raise Error, "Daemon unreachable at #{@base}: #{e.message}. Start it with: srift daemon start"
    end
  end
end

File: sdk/node/srift.mjs
/**
 * SRIFT Node.js SDK — zero-dependency client for the SRIFT local daemon.
 *
 * Runs on: Node 18+, Bun, Deno (via npm: specifier), browsers, edge runtimes
 *          (Vercel, Cloudflare Workers, Netlify Edge, Deno Deploy, Lambda@Edge),
 *          Electron, React Native, Hermes, JerryScript.
 *
 * Zero external deps. Uses only the built-in fetch (Node 18+) + EventSource shim.
 *
 * Usage:
 *   import { Srift } from "./srift.mjs";
 *   const s = new Srift();
 *   const { shareUrl } = await s.quickShare("/abs/path/file.zip");
 */

const DEFAULT_BASE = "http://127.0.0.1:3822";

export class SriftError extends Error {
  constructor(message, code) { super(message); this.name = "SriftError"; this.code = code; }
}

export class Srift {
  constructor({ baseUrl, timeoutMs } = {}) {
    this.baseUrl = (baseUrl || (typeof process !== "undefined" && process.env?.SRIFT_BASE_URL) || DEFAULT_BASE).replace(/\/$/, "");
    this.timeoutMs = timeoutMs ?? 30_000;
  }

  async _call(path, method = "GET", body) {
    const ctrl = typeof AbortController !== "undefined" ? new AbortController() : null;
    const timer = ctrl ? setTimeout(() => ctrl.abort(), this.timeoutMs) : null;
    try {
      const r = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers: { "Content-Type": "application/json" },
        body: body == null ? undefined : JSON.stringify(body),
        signal: ctrl?.signal,
      });
      const text = await r.text();
      const json = text ? JSON.parse(text) : {};
      if (!r.ok) throw new SriftError(json.error || `HTTP ${r.status}`, r.status);
      return json;
    } catch (err) {
      if (err instanceof SriftError) throw err;
      throw new SriftError(`Daemon unreachable at ${this.baseUrl}: ${err.message}. Start it with: srift daemon start`);
    } finally {
      if (timer) clearTimeout(timer);
    }
  }

  // Health
  isAlive() { return this._call("/status").then(() => true, () => false); }
  status() { return this._call("/status"); }
  state() { return this._call("/state"); }

  // Session
  startSession({ name, roomSecret } = {}) {
    return this._call("/session/start", "POST", { sessionName: name, roomSecret });
  }
  joinSession(sessionId, { username, roomSecret } = {}) {
    return this._call("/session/join", "POST", { sessionId, username, roomSecret });
  }
  approveJoin(tempUserId) { return this._call("/session/approve", "POST", { tempUserId }); }
  rejectJoin(tempUserId, reason) { return this._call("/session/reject", "POST", { tempUserId, reason }); }
  kickUser(userId) { return this._call("/session/kick", "POST", { userId }); }
  closeSession() { return this._call("/session/close", "POST"); }

  // Transfer
  sendFile(filePath, { protocol } = {}) { return this._call("/send", "POST", { filePath, protocol }); }
  acceptTransfer(fileId, { saveDir } = {}) { return this._call("/receive", "POST", { fileId, saveDir }); }
  async listTransfers() { return (await this.status()).activeTransfers || []; }
  quickShare(filePath, { sessionName } = {}) {
    return this._call("/quick-share", "POST", { filePath, sessionName });
  }

  // Chat
  sendChat(message) { return this._call("/chat/send", "POST", { message }); }
  chatHistory() { return this._call("/chat/history"); }

  // SSE event stream (async iterator)
  async *streamEvents() {
    const r = await fetch(`${this.baseUrl}/api/v1/monitor/events`, { headers: { Accept: "text/event-stream" } });
    if (!r.ok || !r.body) throw new SriftError(`SSE failed: HTTP ${r.status}`);
    const reader = r.body.getReader();
    const dec = new TextDecoder();
    let buf = "";
    let event = "message";
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buf += dec.decode(value, { stream: true });
      let idx;
      while ((idx = buf.indexOf("\n")) !== -1) {
        const line = buf.slice(0, idx).trimEnd();
        buf = buf.slice(idx + 1);
        if (line.startsWith("event: ")) event = line.slice(7).trim();
        else if (line.startsWith("data: ")) {
          let data;
          try { data = JSON.parse(line.slice(6)); } catch { data = line.slice(6); }
          yield { event, data };
          event = "message";
        }
      }
    }
  }
}

// ─── MCP HTTP transport client ──────────────────────────────────────────────
export class SriftMCP {
  constructor({ baseUrl } = {}) {
    this.baseUrl = (baseUrl || DEFAULT_BASE).replace(/\/$/, "");
    this._id = 0;
  }
  async call(method, params) {
    const body = { jsonrpc: "2.0", id: ++this._id, method, ...(params != null ? { params } : {}) };
    const r = await fetch(`${this.baseUrl}/mcp`, {
      method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(body),
    });
    const payload = await r.json();
    if (payload.error) throw new SriftError(payload.error.message, payload.error.code);
    return payload.result;
  }
  initialize(clientName = "srift-node-sdk") {
    return this.call("initialize", {
      protocolVersion: "2025-06-18",
      capabilities: {},
      clientInfo: { name: clientName, version: "2.0.0" },
    });
  }
  async tools() { return (await this.call("tools/list")).tools; }
  tool(name, args = {}) { return this.call("tools/call", { name, arguments: args }); }
  async resources() { return (await this.call("resources/list")).resources; }
  resource(uri) { return this.call("resources/read", { uri }); }
  async prompts() { return (await this.call("prompts/list")).prompts; }
  prompt(name, args = {}) { return this.call("prompts/get", { name, arguments: args }); }
}

export default Srift;
